### PR TITLE
Another attempt to improve the search performance

### LINKED ISSUE #4239: Improve Search Performance
https://github.com/shilangyu/scoop-search has changed my experience with scoop, the slow search has always been one of my pain points as a user.

I'd be willing to PR improvement to the search engine to incorporate the lessons learnd here if there is interest.

### PR DESCRIPTION
**Pull request by r15ch13**
#### Description
This is another attempt to improve the local search performance.
- Not using the `manifest()` function
- Use only one loop over `$apps`
- Push results to `$list` directly
- Use `[System.IO.File]::ReadAllText()` instead of `Get-Content`

#### Motivation and Context

Closes #4239

Relates to #4818

#### How Has This Been Tested?

```pwsh
Measure-Command { scoop search telegram }
Measure-Command { .\bin\scoop.ps1 search telegram }
```

image

#### Checklist:

- [x] I have read the [Contributing Guide](https://github.com/ScoopInstaller/.github/blob/main/.github/CONTRIBUTING.md).
- [x] I have ensured that I am targeting the `develop` branch.
- [ ] I have updated the documentation accordingly.
- [ ] I have updated the tests accordingly.
- [ ] I have added an entry in the CHANGELOG.

### CODE REVIEW

**Comment by niheaven** [comment]
WIP or Ready for review?

**Comment by rashil2000** [comment]
Some more areas for optimization:
- Since we are listing lots of files in function `apps_in_bucket`, using `[System.IO.Directory]::EnumerateFiles()` instead of `Get-ChildItem` should be much faster ([ref](https://stackoverflow.com/questions/14970692/powershell-io-directory-find-file-types-in-all-subdirectories))
- Since we are deserializing a lot of JSON, using `[System.Text.Json.Utf8JsonReader]` should be faster than `ConvertFrom-Json`
- While searching in remotes, since we are making multiple requests, a .NET method should be faster than `Invoke-WebRequest`
- I'm not sure I understand this, but why do we use `Select-Object -ExpandProperty` instead of just directly accessing the field on lines 69, 90 and 100?

File: libexec/scoop-search.ps1 [hunk 1]
```diff
@@ -37,19 +37,31 @@ function bin_match($manifest, $query) {
...
 
 function search_bucket($bucket, $query) {
-    $apps = apps_in_bucket (Find-BucketDirectory $bucket) | ForEach-Object { @{ name = $_ } }
-
-    if ($query) {
-        $apps = $apps | Where-Object {
-            if ($_.name -match $query) { return $true }
-            $bin = bin_match (manifest $_.name $bucket) $query
+    $bucket_dir = Find-BucketDirectory $bucket
+    $apps = apps_in_bucket $bucket_dir
```
Comment (niheaven) [inline libexec/scoop-search.ps1:41]:
```suggestion
    $apps = Get-ChildItem (Find-BucketDirectory $bucket) -Filter '*.json' -Recurse
```

File: libexec/scoop-search.ps1 [hunk 2]
```diff
@@ -37,19 +37,31 @@ function bin_match($manifest, $query) {
...
+
+    $apps | ForEach-Object {
+        $json = [System.IO.File]::ReadAllText("$bucket_dir\$_.json") | ConvertFrom-Json -ErrorAction Continue
```
Comment (niheaven) [inline libexec/scoop-search.ps1:44]:
```suggestion
        $json = [System.IO.File]::ReadAllText($_.FullName) | ConvertFrom-Json -ErrorAction Continue
```

File: libexec/scoop-search.ps1 [hunk 3]
```diff
@@ -37,19 +37,31 @@ function bin_match($manifest, $query) {
...
+
+        if($app.name -match $query) {
+            $list.Add([PSCustomObject]@{
+                Name = $_
```
Comment (niheaven) [inline libexec/scoop-search.ps1:48]:
```suggestion
                Name = $_.BaseName
```

File: libexec/scoop-search.ps1 [hunk 4]
```diff
@@ -37,19 +37,31 @@ function bin_match($manifest, $query) {
...
+                Source = $bucket
+                Binaries = ""
+            })
+        } else {
+            $bin = bin_match $json $query
             if ($bin) {
-                $_.bin = $bin
-                return $true
+                $list.Add([PSCustomObject]@{
+                    Name = $_
```
Comment (niheaven) [inline libexec/scoop-search.ps1:57]:
```suggestion
                    Name = $_.BaseName
```

**Review by niheaven — CHANGES_REQUESTED** [review]
Just for nested manifests

**Comment by niheaven** [comment]
> Since we are listing lots of files in function apps_in_bucket, using [System.IO.Directory]::EnumerateFiles() instead of Get-ChildItem should be much faster ([ref](https://stackoverflow.com/questions/14970692/powershell-io-directory-find-file-types-in-all-subdirectories))

image

Sorry but I found `Get-ChildItem` is faster than `[System.IO.Directory]::EnumerateFiles()` here.

**Comment by r15ch13** [comment]
> * Since we are listing lots of files in function `apps_in_bucket`, using `[System.IO.Directory]::EnumerateFiles()` instead of `Get-ChildItem` should be much faster ([ref](https://stackoverflow.com/questions/14970692/powershell-io-directory-find-file-types-in-all-subdirectories))
It's only run once for every bucket. Could be a problem if you have 1000 buckets. :smile:
> * Since we are deserializing a lot of JSON, using `[System.Text.Json.Utf8JsonReader]` should be faster than `ConvertFrom-Json`
Is it available in PowerShell?

@niheaven 
We need to take the filesystem cache and the drive type into account here. (In my case SSD)

File: libexec/scoop-search.ps1 [hunk 5]
```diff
@@ -37,19 +37,30 @@ function bin_match($manifest, $query) {
...
-            if ($_.name -match $query) { return $true }
-            $bin = bin_match (manifest $_.name $bucket) $query
+    $apps = Get-ChildItem (Find-BucketDirectory $bucket) -Filter '*.json' -Recurse
+
+    $apps | ForEach-Object {
+        $json = [System.IO.File]::ReadAllText($_.FullName) | ConvertFrom-Json -ErrorAction Continue
+
+        if($app.name -match $query) {
+            $list.Add([PSCustomObject]@{
+                Name = $_.BaseName
```
Comment (niheaven) [inline libexec/scoop-search.ps1:47]:
```suggestion
        $json = [System.IO.File]::ReadAllText($_.FullName) | ConvertFrom-Json -ErrorAction Continue
        $name = $_.BaseName

        if ($name -match $query) {
            $list.Add([PSCustomObject]@{
                Name = $name
```

File: libexec/scoop-search.ps1 [hunk 6]
```diff
@@ -37,19 +37,30 @@ function bin_match($manifest, $query) {
...
+                Source = $bucket
+                Binaries = ""
+            })
+        } else {
+            $bin = bin_match $json $query
             if ($bin) {
-                $_.bin = $bin
-                return $true
+                $list.Add([PSCustomObject]@{
+                    Name = $_.BaseName
```
Comment (niheaven) [inline libexec/scoop-search.ps1:56]:
```suggestion
                    Name = $name
```

**Review by niheaven — CHANGES_REQUESTED** [review]

File: libexec/scoop-search.ps1 (hunk 5 above)
Comment (niheaven) [reply libexec/scoop-search.ps1:47]: We don't have a variable called `$app`, so all search fallback to bin search.

image

Comment (r15ch13) [reply libexec/scoop-search.ps1]: yep, missed that one :smile:

**Comment by niheaven** [comment]
That's LGTM, and it needs a updated changelog :).

**Comment by rashil2000** [comment]
> Is it available in PowerShell?

Yes

image

**Comment by rashil2000** [comment]
> Sorry but I found Get-ChildItem is faster than [System.IO.Directory]::EnumerateFiles() here.

On multiple repeated runs (i.e. warm cache), EnumerateFiles is almost always twice as fast as Get-ChildItem in my testing.

**Comment by niheaven** [comment]
image

https://learn.microsoft.com/en-us/dotnet/api/system.text.json.jsondocument.parse?view=net-7.0

Not available for .NET 4.5

**Comment by r15ch13** [comment]
Oh, I was searching for `[System.Text.Json.Utf8JsonReader]` explicitly... :grin:

**Comment by niheaven** [comment]
> On multiple repeated runs (i.e. warm cache), EnumerateFiles is almost always twice as fast as Get-ChildItem in my testing.

If we just run `Get-ChildItem` once for one bucket, that's not worthwhile to use `EnumerateFiles`. And the output of `EnumerateFiles` is `string[]` which doesn't have `FullName` or `BaseName`.

**Comment by niheaven** [comment]
Out of topic here, the minimal requirements of Scoop are WindowsPowerShell and .Net Framework 4.5 which really lacks some new features of modern PowerShell and .Net.

Could this be updated? (To newest PowerShell)

**Comment by rashil2000** [comment]
> * I'm not sure I understand this, but why do we use `Select-Object -ExpandProperty` instead of just directly accessing the field on lines 69, 90 and 100?

Alright, what about this one?

**Comment by r15ch13** [comment]
It goes down to around 700ms from 1.3sec by using `JsonDocument.Parse()`. My C# implementation runs at around 300ms.

**Comment by rashil2000** [comment]
Since `[System.Text.Json]` is unavailable in PS 5.1, there's another way: we could reuse the NewtonSoft.Json DLL. As mentioned in this [SO post](https://stackoverflow.com/a/58169326), it is also much faster than `ConvertFrom-Json`.

**Comment by niheaven** [comment]
Too complicated for the JSON parsing function here... And `[System.Text.Json]` is not available for .Net Frameworks (4.5, 5 and all version before .Net Core)

The loop and generic list (this is not needed even, IMO. I've tried to replace all array '@()' to generic list but the time savings is 😭) are enough for `scoop search`, and if `ConvertFrom-Json` matters, let's create a new PR.

**Comment by r15ch13** [comment]
Added a fallback if `System.Text.Json` is not available

**Comment by niheaven** [comment]
I still don't think we need to speed up `ConvertFrom-Json`. From tens of seconds to seconds is important, but from seconds to milliseconds is not.

If this is needed indeed, tweak (or migrate to) the `parse_json()` function is better imo.

**Comment by r15ch13** [comment]
Performance right after starting the PC still sucks.

image

**Comment by niheaven** [comment]
The two different searching time are all using this PR's solution? Wow...

**Comment by r15ch13** [comment]
Yes, just ran the same code twice after restarting the PC.

File: libexec/scoop-search.ps1 [hunk 7]
```diff
@@ -11,6 +11,11 @@ param($query)
 
 $list = [System.Collections.Generic.List[PSCustomObject]]::new()
 
+if(!$query -or $query -in @($null, '-h', '--help', '/?')) {
```
Comment (rashil2000) [inline libexec/scoop-search.ps1:14]: Uh.. this is a breaking change 😅

`scoop search` without arguments has been used to list all local manifests, facilitating commands like `scoop search | ? source -eq extras | measure`.

Comment (r15ch13) [reply libexec/scoop-search.ps1]: Oh, yeah for pipelines. okay reverted 😄

**Comment by justrainer** [comment]
Hello. Whats the status of this? From reading the messages here, it seems to me all painpoints have been agreed upon, no? As someone coming back to Windows from Arch, I find it super painful to have to wait multiple seconds for every search I do, and I'd much appreciate any improvement to the search speed, no matter how big or small. Just wanted to add my two cents - thanks for all the hard work!

**Comment by JVimes** [comment]
Really excited about this one 🙂 Is it still on the radar?

**Comment by r15ch13** [comment]
Messed up the branch. Created a new PR https://github.com/ScoopInstaller/Scoop/pull/5644

**Pull request closed without merging**